File: language/src/main/java/brj/nodes/builtins/PrStrNode.java
package brj.nodes.builtins;

import brj.BridjeContext;
import brj.BridjeLanguage;
import brj.nodes.ExprNode;
import com.oracle.truffle.api.dsl.CachedContext;
import com.oracle.truffle.api.dsl.CachedLanguage;
import com.oracle.truffle.api.dsl.Specialization;
import com.oracle.truffle.api.frame.VirtualFrame;
import com.oracle.truffle.api.interop.InteropLibrary;
import com.oracle.truffle.api.interop.UnsupportedMessageException;
import com.oracle.truffle.api.nodes.UnexpectedResultException;

import static brj.BridjeTypesGen.expectString;

public abstract class PrStrNode extends ExprNode {

    private final InteropLibrary objLib = InteropLibrary.getUncached();

    public PrStrNode(BridjeLanguage lang) {
        super(lang);
    }

    @Specialization
    public String doExecute(VirtualFrame frame,
                            @CachedLanguage BridjeLanguage lang,
                            @CachedContext(BridjeLanguage.class) BridjeContext ctx) {
        try {
            Object obj = frame.getArguments()[1];
            if (!objLib.hasLanguage(obj) || objLib.getLanguage(obj) != BridjeLanguage.class) {
                obj = lang.getLanguageView(ctx, obj);
            }
            return expectString(objLib.toDisplayString(obj));
        } catch (UnexpectedResultException | UnsupportedMessageException e) {
            throw new RuntimeException(e);
        }
    }

}

File: language/src/main/kotlin/brj/builtins/PrStrNode.kt
package brj.builtins

import brj.BridjeLanguage
import brj.BridjeTypesGen.expectString
import com.oracle.truffle.api.TruffleLanguage.ContextReference
import com.oracle.truffle.api.TruffleLanguage.LanguageReference
import com.oracle.truffle.api.dsl.Specialization
import com.oracle.truffle.api.interop.InteropLibrary

private val LANG_REF = LanguageReference.create(BridjeLanguage::class.java)
private val CTX_REF = ContextReference.create(BridjeLanguage::class.java)

@BuiltIn("pr-str")
abstract class PrStrNode(lang: BridjeLanguage) : BuiltInFn(lang) {
    @field:Child
    private var objLib = InteropLibrary.getFactory().createDispatched(3)

    @Specialization
    fun doExecute(obj: Any): String {
        val objView =
            if (!objLib.hasLanguage(obj) || objLib.getLanguage(obj) != BridjeLanguage::class.java) {
                LANG_REF.get(this).getLanguageView(CTX_REF.get(this), obj)
            } else obj

        return expectString(objLib.toDisplayString(objView))
    }
}